### PR TITLE
SALTO-4458 improve path index hints generation consistency

### PR DESCRIPTION
**Pull request by Uri-Bar**
SALTO-4458 improve path index hints generation consistency
we added for keys to the pathIndex such as .annotation .field and .attr elemId (e.g: salto.obj.field).
now when the field of an object are divided between two files `salto.obj.field`   will point to both of them.
And if the few annotation under the same file, we wont have a key of each annotation in the pathIndex. instead `salto.obj.annotation` will point at the relevant file.

---
_Release Notes_: 
none

---
_User Notifications_: 
none

### CODE REVIEW

**Comment by ori-moisis** [comment]
does this make any effect on users?
not sure I understand the release notes...

File: packages/workspace/src/workspace/path_index.ts [hunk 1]
```diff
@@ -69,28 +71,44 @@ const getAnnotationTypesPathHints = (
   const fragmentsWithNonEmptyAnnoTypes = fragments.filter(
     f => !_.isEmpty(f.value.annotationRefTypes)
   )
-
+  if (_.isEmpty(fragmentsWithNonEmptyAnnoTypes)) {
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/path_index.ts:74]: can we reduce the use of double negatives here by renaming `fragmentsWithNonEmptyAnnoTypes` to just `fragmentsWithAnnoTypes`

asking 'isEmpty` on something with "NonEmpty" just seems weird

File: packages/workspace/src/workspace/path_index.ts [hunk 2]
```diff
@@ -154,19 +171,22 @@ const getElementPathHints = (
   const valueHints = elementFragments.every(f => isInstanceElement(f.value))
     ? getValuePathHints(
       (elementFragments as Fragment<InstanceElement>[])
-        .map(f => ({ value: f.value.value, path: f.path })),
+        .map(f => ({ value: _.merge({}, f.value.value, f.value.annotations), path: f.path })),
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/path_index.ts:174]: similar note about needing a comment to explain what we are doing with instance elements and why.
maybe we should couple the two conditions so we only have to explain this once?

File: packages/workspace/src/workspace/path_index.ts [hunk 3]
```diff
@@ -69,28 +71,44 @@ const getAnnotationTypesPathHints = (
...
 
 const getAnnotationPathHints = (
   fragments: Fragment<Element>[],
 ): PathHint[] => {
   const elem = fragments[0].value
+  const fragmentsWithAttr = fragments.filter(f => !_.isEmpty(f.value.annotations))
+  if (_.isEmpty(fragmentsWithAttr)) {
+    return []
+  }
+  if (isInstanceElement(elem)) {
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/workspace/src/workspace/path_index.ts:100]: ack

File: packages/workspace/src/workspace/path_index.ts (hunk 1 above)
Comment (Uri-Bar) [reply packages/workspace/src/workspace/path_index.ts:74]: ack

File: packages/workspace/src/workspace/path_index.ts [hunk 4]
```diff
@@ -69,28 +71,44 @@ const getAnnotationTypesPathHints = (
...
+    return []
+  }
+  if (fragmentsWithAttr.length === 1) {
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/workspace/src/workspace/path_index.ts:103]: ack will remove and test it

File: packages/workspace/src/workspace/path_index.ts [hunk 5]
```diff
@@ -43,97 +43,99 @@ export type PathIndex = RemoteMap<Path[]>
...
   // We only have 3 cases to handle: Object type (which can be split among files)
   // or a single list/primitive value.
+  const valueTopLevelKey = [{
+    key: elemID.getFullName(),
+    value: makeArray(fragments.map(f => f.path)),
+  }]
+  if (fragments.length === 0) {
+    // No fragments with this ID, so no path hints needed
+    return []
+  }
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/path_index.ts:53]: nit - move this the be before creating the `valueTopLevelKey`

File: packages/workspace/test/workspace/path_index.test.ts [hunk 6]
```diff
@@ -536,3 +536,238 @@ describe('updatePathIndex', () => {
...
+    expect(hintsKeys).toContain('salto.obj')
+    const fieldHints = pathHints.filter(p => p.key.includes('field'))
+    expect(hintsKeys).toContain('salto.obj.field')
+    expect(fieldHints.length).toEqual(1)
+    expect(fieldHints[0].value.length).toEqual(1)
+    const annoHints = pathHints.filter(p => p.key.includes('annotation'))
+    expect(hintsKeys).toContain('salto.obj.annotation')
+    expect(annoHints.length).toEqual(1)
+    expect(annoHints[0].value.length).toEqual(1)
+    expect(hintsKeys).toContain('salto.obj.attr')
```
Comment (ori-moisis) [inline packages/workspace/test/workspace/path_index.test.ts:589]: I think it is a bit hard to follow the `expect`s here
I get lost with all the "length" checks, it also doesn't actually validate we get the correct paths.

two options I see here:
1. Turn this into an object and then check for each key and value (or all together), e.g `const pathHintObj = { ...pathHints.map(hint => ({ [hint.key]: hint.value }) }` and then `expect(pathHintObj).toHaveProperty('salto.obj', [...])`
2. Use `expect(pathHints).toContain(expect.objectMatching(...))` with the specific paths we are looking for

File: packages/workspace/test/workspace/path_index.test.ts [hunk 7]
```diff
@@ -536,3 +536,227 @@ describe('updatePathIndex', () => {
...
+      annotationRefsOrTypes: {
+        anno: createRefToElmWithValue(BuiltinTypes.STRING),
+      },
+      annotations: {
+        anno: 'Bye',
+      },
+      path: ['salto', 'obj', 'annotations'],
+    })
+    const pathHints = getElementsPathHints([singleFieldObj, singleFieldObjAnnotations])
+    expect(pathHints.length).toEqual(4)
```
Comment (ori-moisis) [inline packages/workspace/test/workspace/path_index.test.ts:579]: I think we don't need this check because this should be covered by the "toEqual" check below IIUC, but for future reference, when asserting the length of an array:
```suggestion
    expect(pathHints).toHaveLength(4)
```

File: packages/workspace/test/workspace/path_index.test.ts [hunk 8]
```diff
@@ -536,3 +536,227 @@ describe('updatePathIndex', () => {
...
+        ping: 'pong',
+      },
+      path: ['salto', 'obj', 'annotationsTwo'],
+    })
+    const pathHints = getElementsPathHints([
+      objFragAnnotationsOne, objFragAnnotationsTwo, objFragStdFields, objFragCustomFields,
+    ])
+    expect(pathHints.length).toEqual(10)
+    expect(pathHints).toEqual(expect.arrayContaining([
+      expect.objectContaining({ key: 'salto.obj.annotation', value: [['salto', 'obj', 'annotationsOne'], ['salto', 'obj', 'annotationsTwo']] }),
```
Comment (ori-moisis) [inline packages/workspace/test/workspace/path_index.test.ts:638]: I think extracting these paths to variables would make the test more readable

e.g.
```suggestion
      expect.objectContaining({ key: 'salto.obj.annotation', value: [annotationsOnePath, annotationsTwoPath] }),
```

**Review by ori-moisis — APPROVED** [review]

**Comment by ori-moisis** [comment]
I think we should have more information about what this PR does in the PR description.
the description currently uses very generic words...
